### PR TITLE
DatabaseDeveloperPageExceptionFilter runs next filter if handling fails

### LINKED ISSUE #32076: AddDatabaseDeveloperPageExceptionFilter seems broken
## File a bug
I created a starndard asp.net core 5.0 razor page app. I then edited the connectionstring to purposely make it an invalid connectionstring. When I checked the errors it showed the following:

```
fail: Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseDeveloperPageExceptionFilter[7]
      An exception occurred while calculating the database error page content. Skipping display of the database error page.
```
In firefox I end up seeing a blank page, in Edge I see error 500 but not the standard developer exception page I've come to expect

I'm not sure if it's related to dotnet/efcore#24648 

### Include your code
Start up code
```C#
public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddDbContext<ApplicationDbContext>(options =>
                options.UseSqlServer(
                    Configuration.GetConnectionString("DefaultConnection")));
            services.AddDatabaseDeveloperPageExceptionFilter();

            services.AddDefaultIdentity<IdentityUser>(options => options.SignIn.RequireConfirmedAccount = true)
                .AddEntityFrameworkStores<ApplicationDbContext>();

            // ...
            services.Configure<DataProtectionTokenProviderOptions>(options =>
            {
                options.TokenLifespan = TimeSpan.FromSeconds(20); // Sets the expiry to two days
            });


            services.AddControllersWithViews();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseMigrationsEndPoint();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                app.UseHsts();
            }
            app.UseHttpsRedirection();
            app.UseStaticFiles();

            app.UseRouting();

            app.UseAuthentication();
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                    name: "default",
                    pattern: "{controller=Home}/{action=Index}/{id?}");
                endpoints.MapRazorPages();
            });
        }
    }
```

### Include stack traces / Include verbose output

```
at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.HttpContextDatabaseContextDetailsExtensions.GetContextDetailsAsync(HttpContext httpContext, Type dbcontextType, ILogger logger)
         at Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseDeveloperPageExceptionFilter.HandleExceptionAsync(ErrorContext errorContext, Func`2 next)
      ClientConnectionId:00000000-0000-0000-0000-000000000000
      Error Number:-1983577849,State:0,Class:20
fail: Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseDeveloperPageExceptionFilter[7]
      An exception occurred while calculating the database error page content. Skipping display of the database error page.
      Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: SNI_PN11, error: 0 -  [x89C50107])
       ---> System.ComponentModel.Win32Exception (0x89C50107): Unknown error (0x89c50107)
         at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
```

Appsettings
```json
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=(localdb)\\mssqllocaldbMakeThisStringNotConnectable;Database=aspnet-WebApplication45-97C4AE3B-1D8B-4A59-A330-C56E58400F49;Trusted_Connection=True;MultipleActiveResultSets=true"
  },
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*"
}

```

### Include provider and version information

EF Core version: 5.0.5
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 5.0
Operating system: Windows
IDE: Visual Studio 2019 16.9.4

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Fixes https://github.com/dotnet/aspnetcore/issues/32076.

### CODE REVIEW

File: src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs [hunk 1]
```diff
@@ -85,8 +85,10 @@ public async Task HandleExceptionAsync(ErrorContext errorContext, Func<ErrorCont
             catch (Exception e)
             {
                 _logger.DatabaseErrorPageMiddlewareException(e);
-                return;
             }
+
+            // Error could not be handled, let the next filter try
+            await next(errorContext);
```
Comment (Tratcher) [inline src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs:91]: When the exception occurred matters. The following are checked before calling the handers at all:
https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddleware.cs#L100-L109
If one handler is going to fall back to another, it should do similar steps before calling next.

File: src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs [hunk 2]
```diff
@@ -47,6 +48,7 @@ public async Task HandleExceptionAsync(ErrorContext errorContext, Func<ErrorCont
             if (!(errorContext.Exception is DbException))
             {
                 await next(errorContext);
+                return;
```
Comment (JunTaoLuo) [inline src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs:51]: Unit tests caught this bug.

File: src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs:104]: Hmm now I think about it shouldn't we check for Response.HasStarted at the beginning of the filter as well?

Comment (Tratcher) [reply src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs:104]: It's checked before the filter is called. If we follow the middleware pattern, it's each filters responsibility to only call next if the request is still viable.

Comment (JunTaoLuo) [reply src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseDeveloperPageExceptionFilter.cs:104]: I suspect this is something that's easily missed, maybe we should be enforcing this somehow.

File: src/Middleware/Diagnostics.EntityFrameworkCore/src/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs [hunk 3]
```diff
@@ -75,6 +75,9 @@ internal static class DiagnosticsEntityFrameworkCoreLoggerExtensions
             new EventId(7, "DatabaseErrorPageException"),
             "An exception occurred while calculating the database error page content. Skipping display of the database error page.");
 
+        private static readonly Action<ILogger, Exception?> _responseStartedDatabaseDeveloperPageExceptionFilter =
+            LoggerMessage.Define(LogLevel.Warning, new EventId(1, "ResponseStarted"), "The response has already started, the database developer page exception filter will not be executed.");
```
Comment (Tratcher) [inline src/Middleware/Diagnostics.EntityFrameworkCore/src/DiagnosticsEntityFrameworkCoreLoggerExtensions.cs:79]: The DB filter already failed, it's the next filter that won't be executed, right?

**Review by Tratcher — APPROVED** [review]

**Pull request merged**